Jetsnack/app/src/main/java/com/example/jetsnack/ui/JetsnackApp.kt

```diff
@@ -17,30 +17,193 @@
+class ScaffoldStateHolder(
+    val navController: NavHostController,
+    private val snackbarHostState: SnackbarHostState,
+    coroutineScope: CoroutineScope,
+    lifecycle: Lifecycle,
+    initialSnackbarMessages: List<String>
+) {
+    // Tabs for JetsnackBottomBar
+    val tabs = HomeSections.values()
+
+    // Queue a maximum of 3 snackbar messages
+    private val snackbarMessages = StateChannel<String>(3, BufferOverflow.DROP_OLDEST)
+
+    init {
+        coroutineScope.launch {
+            // Restore snackbar messages state
+            initialSnackbarMessages.forEach { showSnackbar(it) }
+
+            // Process snackbar events only when the lifecycle is at least STARTED
+            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                snackbarMessages.stream
+                    .collect { message ->
+                        snackbarHostState.showSnackbar(message)
+                    }
+            }
+        }
+    }
+
+    suspend fun showSnackbar(message: String) {
+        snackbarMessages.send(message)
+    }
+
+    fun navigateToSnackDetail(from: NavBackStackEntry, snackId: Long) {
+        // In order to discard duplicated navigation events, we check the Lifecycle
+        if (from.lifecycleIsResumed()) {
+            navController.navigate("${MainDestinations.SNACK_DETAIL_ROUTE}/$snackId")
+        }
+    }
+
+    fun onBackPress() {
+        navController.navigateUp()
+    }
+
+    companion object {
+        fun ScaffoldStateHolderSaver(
+            navController: NavHostController,
+            snackbarHostState: SnackbarHostState,
+            coroutineScope: CoroutineScope,
+            lifecycle: Lifecycle
+        ) = listSaver<ScaffoldStateHolder, String>(
+            save = { it.snackbarMessages.pendingElements },
+            restore = { pendingMessages ->
+                ScaffoldStateHolder(
+                    navController, snackbarHostState, coroutineScope, lifecycle, pendingMessages
+                )
+            }
+        )
+    }
+}
+
+/**
+ * Implementation of a Channel whose buffered elements can be accessed synchronously so that they
+ * can be saved/restored on activity and process recreation.
+ *
+ * In this implementation, the exposed `stream` flow is supposed to be collected once at a time
+ * since in case there are multiple collectors, only one will get the event.
+ * If events need to be broadcasted to multiple collectors, expose stream using the `.shareIn`
+ * operator instead.
+ */
+class StateChannel<T>(
+    private val capacity: Int,
+    private val onBufferOverflow: BufferOverflow = BufferOverflow.SUSPEND
+) {
+    // Cache of the Channel buffer used to restore state
+    val pendingElements = mutableStateListOf<T>()
+    private val pendingElementsMutex = Mutex()
+
+    private val _channel = Channel<T>(capacity, onBufferOverflow)
+    val stream = _channel
+        .receiveAsFlow()
+        .onEach { pendingElementsMutex.withLock { pendingElements.remove(it) } }
```

Comment:
How do you feel about replacing this with `removeFirst()`, instead of `remove(it)`?
If `removeFirst()` or `remove(it)` ever do something different, then I think `remove(it)` could have unintended consequences.

Reply:
+1, definitely!